### ISSUE TITLE
RFC - Boilerplate presets

### ISSUE THREAD

**Issue by Vadorequest** [UnlyEd/next-right-now#18]
# Context

This boilerplate has been built based on a real professional use-case. 

**This use case relies on the following built-in 3rd parties:**
- Monitoring (Sentry 3rd party, large freemium)
- Analytics (Amplitude 3rd party, large freemium)
- GraphQL API (GraphCMS 3rd party, small freemium)
- I18n (Locize 3rd party, 2w free trial, **not free**)
- Hosting (Zeit 3rd party, large freemium, no pay-or-leave)

**Also, it features the following built-in design choices:**
- SSR (no SSG nor static optimization are possible without altering the boilerplate in a non-straightforward way)
- Multi-tenants/monorepo (not single-tenant friendly without altering the boilerplate in a non-straightforward way)

**Legend:**
Freemium means free with "more or less" limited capabilities on free plan. 
- Small means the free plan is considered too limited for most real use cases
- Large means the free plan is limited but large enough to cover non enterprise-grade needs 

# Concerns
There are 2 main concerns at hand.

## It's too complicated/opinionated
Opinions can be good, I personally don't regret those 3rd party choices. But not everyone like them, not everyone need them either. They are currently **built-in**, but it would be better to have them **opt-in** and let people easily choose what boilerplate preset fits their need the most closely.

## It's not free
Locize provides a free trial, but it's just a trial and it's not free. Thus, it's a strong issue for associations and organisation that don't have any funds, or can't afford it.

GraphCMS free plan is limited to 500 records, which can be considered "good enough" or "far too little" depending on the use cases.

# Needs
- Quickly understanding which boilerplate to use based on my current requirements
- Quickly clone the boilerplate I want to use (git clone specific branch)
- Quickly see what changes have been applied to add/remove a feature compared to another preset
- Ability to apply identical updates to multiple presets easily (ease to maintain all those presets by reducing issues when applying changes that are common to all, such as dependency updates, security updates, good practices, etc.) 

# Solution
I believe the best way to keep the current advanced/opinionated boilerplate while allowing for simpler/faster re-use is to create "presets" from the initial boilerplate. A preset could be considered as similar to a fork, but lives within the same repository.

**After much thinking on the matter, I've come to the following requirements:**
- A preset should live within a git branch. 
    - Each of those branch will have its own "preset origin" from where it got "forked".
    - Those branch must not be confused with working branches, it must be obvious they are official, stable presets where anyone can get started from.
- Each preset should specify whether it is using SSR or SSG design
- Each preset should specify whether it is using multi-tenants or single-tenant design
- Each preset should specify which 3rd parties or features have been removed (compared to the most complex preset, which currently lives in the `master` branch)
- This boilerplate will likely evolve in the future, maybe more features or 3rd parties will be added/removed/changed. It should be made easy to create major versions of the presets, based on a different base major preset.
- Each preset must be documented in the main README.md file, and explained thoroughly. A link to the branch and to the PR must be present (against its "preset origin" so that only minimal changes are shown in the diff), also a `git clone` command must be present to allow for quick local clone of a preset 
- A PR can be created (but closed to avoid false-positive PR) for quick review of source code changes between two presets. It will also allow the community to post comment on the PR source code changes if questions arise and will reduce the github issues created if proper answers are given to those questions, as future readers may also ask themselves the same question and see a proper answer right there. 
- A preset cannot contain uppercased letters, as Zeit won't allow to create a project with uppercases as per now@17 CLI version

**Which led to the following specification:**
- **Branch naming**, each official preset's branch must (in order):
    - Start with `v` (for "version", as Next.js and Zeit may evolve in the future, and require to provide different presets (old ones and new ones) while keeping the same features and vendors)
    - Then with the major version of the preset (i.e: `1`)
    - Then with a `ssr`/`ssg` tag
    - Then with a `mst` tag (if multiple single-tenants design), or `mt` tag if multi-tenant (single-tenancy isn't something we aim to support, because MST is better (ROI), in our opinion)
    - Then, with a shortened list of all features/vendor differences with the "origin preset" (we can't list them all due to domain name characters limit (53 max))
        - e.g: `v1-ssr-mst-aptd-gcms-lcz-sty` ([see full list](https://unlyed.github.io/next-right-now/getting-started/select-preset))
- **Zeit project naming**, each official preset's`now.json` file must (in order) :
    - Start with `nrn-`
    - Then with the branch name
    - Additionally, for `mst` branches, they will also specify "customer1" or customer2" at the end of the name

### Most complete example, with all 3rd parties, SSR and monorepo design
(basically what's currently in the master branch)

Branch name: `v1-ssr-mst-aptd-gcms-lcz-sty`
Zeit project's names: `nrn-v1-ssr-mst-aptd-gcms-lcz-sty-c1`, `nrn-v1-ssr-mst-aptd-gcms-lcz-sty-c2` (in now.*.json files, as `name`)

### Example without monorepo design (single-tenant instead)
Branch name: `v1-ssr-mt-aptd-gcms-lcz-sty` 
Zeit project's names: `nrn-v1-ssr-mt-aptd-gcms-lcz-sty`
From: `v1-ssr-mst-aptd-gcms-lcz-sty` branch

### Example without Locize 3rd party
Branch name: `v1-ssr-mst-aptd-gcms-sty`
Zeit project's names:  `v1-ssr-mst-aptd-gcms-sty`
From: `v1-ssr-mst-aptd-gcms-lcz-sty` branch

### Example without i18n nor monorepo
Branch name: `v1-ssr-mt-aptd-gcms-sty` 
Zeit project's names: `nrn-v1-ssr-mt-aptd-gcms-sty`
From: `v1-ssr-mt-aptd-gcms-lcz-sty` branch (because it's the closest of what we need)

## Needs checkup
- [x] Quickly understanding which boilerplate to use based on my current requirements
    - The README will help users finding the most appropriate preset of the boilerplate that best fit their need, and will allow them to quickly clone the preset locally and see source code changes through PR diff
- [x] Quickly clone the boilerplate I want to use (git clone specific branch)
    - The specific git clone command will be included in the preset documentation in the README
- [x] Quickly see what changes have been applied to add/remove a feature compared to another preset
    - The source code changes will be available through the preset PR diff view, people will also have the ability to comment things they don't understand or don't agree with (community feedback)
- [x] Ability to apply identical updates to multiple presets easily (ease to maintain all those presets by reducing issues when applying changes that are common to all, such as dependency updates, security updates, good practices, etc.) 
    - Possible through cherry pick or git patches, straightforward as long as there are no conflict.

# Feedback
Please share your questions, concerns or suggestion.

**Comment by Vadorequest** [comment]
The tag `monorepo` could be replaced by `mr`/`mono` (MonoRepo) or `mst` (Multiple Single-Tenant), I'm thinking `mst` is the best choice between both as it's the closest to what actually happens under the hood.

**Comment by jajourda** [comment]
# Thoughts from a Noob

@Vadorequest , what one person calls "_opinionated_" in this context, I am choosing to call **_didactic_**. For starters, a quick definition for shared terminology:

> **noob** for the duration of this rant, the noun is an umbrella term describing any person who is loosely or rigidly in the mode of _guided, exploratory learning_.

![image](https://user-images.githubusercontent.com/1839603/77523875-ce92df80-6e5c-11ea-8d09-0bb770c743ab.png)

## You work for UnlyEd ... 
I saw this tag at the end of next-right-now's readme:

> #TECHFORGOOD #EDUCATIONFORALL

### _If this mentality is for real, noobs should be welcomed_
I find this principle at work: repos that people deem "un-opinionated" tend to increase the suffering of noobs like myself. To be un-opinionated is to be generic; to be generic is not overcomplicate. This means all the examples provided by said "universal, un-opinionated" repos are a mile wide and an inch deep. As soon as a noob tries to clone the repo and get started... EXPLOSION - I have a peculiar idea/need/concern that these generic to-do app docs didn't/can't/won't cover.

## Enter the Didactic E2E approach
I , as a noob, (and I suspect many others) are in need of a [pedagogue](https://www.jstor.org/stable/3121511?seq=1), not simply another tool. An opinionated framework is like being trained by a superior I'm supposed to imitate. 

### This approach has its own frustrations:

But I prefer to pick my problems. I pick this poison. Best practices are ignored when noobs wild-west and cobble together apps. I prefer to learn through imitating the right way if it's made available to me.

![image](https://user-images.githubusercontent.com/1839603/77527293-46afd400-6e62-11ea-8caf-21a74309bc89.png)

## Concerns

### 1.  opinionated

> It's too complicated/opinionated

#### _This should be a downgraded concern_

While not to be ignored, providing a **an already functioning path** is better for a noob than **no path at all** (i.e. figuring it out on our own). I think this can be overcome in a number of ways:

- Really clear documentation (you are well on the way to this and will likely need help)
- Well annotated code (I forked, configured and pushed the repo to zeit yesterday afternoon, and in the process **was thoroughly impressed** with how frequently your code had annotated comments that explained the method behind the madness.


### 2. free

> It's not free

#### _This is where I think most obstacles to adoption will arise (**ESPECIALLY AMONG NOOBS**)_

##### old to you, new to many/most
Your practices are established for you, but will likely be varying shades of _new_  for others. 

##### different approaches may need to avoid payment as long as possible or altogether
People who are in the mode of being guided by you into new/different practices are not likely going to be immediately turning their work into a codebase that can financially support itself (freemium time frames are too tight, e.g. locize). Some will want to adapt your practices at a much slower pace. 

##### different use cases : untraditional revenue streams instead of traditional customer as client
Some may want to actually try to build a test that is **actually to scale**, but will be prohibited from doing so by the low ceilings imposed by freemium plans (I myself want to use your approach for a content sharing platform that will have thousands of content entries, but will be dependent on a  revenue model that is based on voluntary monetization [e.g. patreon]. I am actually going to need some time and space to create loads of rich content that will quickly hit GraphCMS's ceiling before I can even gain steam).

## Needs
I am fully in agreement with these needs. I would add some more potential needs:

1. Some proposed, enumerated use cases as examples for each variation of the codebase.
2. A beefing up of hyperlinks in the readmes to expand the breadth of help offered for noobs (i.e. hyperlinking conceptual terms or broader practices). 
3. A supporting docs website that can easily be updated with **supplemental** material that will help empower noobs to know which variant is the right choice and why.

## Solution
This is where you being opinionated is great: you're being didactic, so you're the teacher. Picking the solution will be your job @Vadorequest  😃 .

### Noob concern : naming convention scares me
Consider tightly coupling simple, clear, readable thorough documentation to the appropriate branches. I fear having to do research just to understand the name of a branch in the future.
![image](https://user-images.githubusercontent.com/1839603/77548252-ad45e980-6e84-11ea-92d7-652522675188.png)

**Comment by Vadorequest** [comment]
I personally dislike the "noob" naming, as it is usually referred to with a negative intent. I prefer to it the term "beginner" which doesn't bear such negativity.

I understand your point regarding how a transparent, documented and comprehensible strong opinion is favorable for beginners compared to more "flex" opinions, which often lead to getting lost due to too many choices/possibilities. I personally second that.

It's probably better for beginners, I wonder what experimented developer think on that matter.

I like the term "didactic" compared to "opinionated", it's also closer to the truth and the goal of this repository. I'm by no mean a teacher but I do my best to educate and document my choices and design for others to understand my motivations easily.

We both agree about the boilerplate providing at least one completely free ($) variant.

Agreed about documenting each variant so that it's easy to tell which one to use based on one's needs.

Adding a "resources" or "reference" section for external links and article may be interesting. You just went through the whole documentation, what would you have liked to find there? What was missing?

I'm not familiar with github documentation websites, never used it. Also, I'm not sure I'd have the time to build such a documentation, it's a lot of work. But a community solution is welcome.

Regarding the branch naming, it's a bit complex at first, but does make sense as soon as you're familiar with the tags. (ssr/ssg, i18n, locize, sentry, etc.)
The goal isn't for newcomers to go through the branches directly, but rather to use the dedicated section in the README to figure out which variant is best for them and how it compares to other.

**Comment by jajourda** [comment]
# A Beginner speaks
@Vadorequest , I retract my ~~noob~~ language! I am a **BEGINNER**.

## github pages would be great
I do have experience with github pages and jekyll. If you are open to it, I would consider helping with such an endeavor. 

## love your faq
I like the idea of the docs being an faq that has actual _gotchas_ and problems that have been run into (similar to the faq section you added to the readme).

## the documentation

> Adding a "resources" or "reference" section for external links and article may be interesting. You just went through the whole documentation, what would you have liked to find there? What was missing?

Honestly, I kind of wish the how to use guide was front and center on the landing readme. It would be great to attempt something that is step-by-step to get the entire project off the ground in order. If you want to renew your goals around the introduced term "didactic" then don't give **any** bullet points that are agnostic of steps. 

### For Example, Where SHOULD I start
It was quite messy trying to determine when and in what specific context to apply _tips_ and which step to do first.
![image](https://user-images.githubusercontent.com/1839603/77567167-25b8a480-6e9d-11ea-9506-f707e6e0cac3.png)


### No Criticism: Just Keep Faithfully giving TLC to the Readmes
 Just tell them in numbered steps what to do and allow them to skip the step with footnoted info if applicable.

**Comment by Vadorequest** [comment]
> I do have experience with github pages and jekyll. If you are open to it, I would consider helping with such an endeavor.

Totally, if you could setup something similar I'd be happy to make use of it. I'd also love to know what steps you've followed to do so, so that I can learn from you as well ;)

Let me know what you'd need to get started. Maybe I need to create a new repository?
If you have any example of what you intend to build, that'd be awesome. Also, I believe we should move this discussion to a dedicated issue.

> If you want to renew your goals around the introduced term "didactic" then don't give any bullet points that are agnostic of steps.

I'm having a hard time understanding what you really mean by that, could you rephrase please? :)

And don't worry about criticism, it's much welcome as long as it's constructive and not aimed at someone in particular. 😉

**Comment by Vadorequest** [comment]
PR https://github.com/UnlyEd/next-right-now/pull/22 is implementing this specification.

**Comment by Vadorequest** [comment]
I'm thinking about renaming this feature from "variant" to "preset". I believe people are more familiar with the later.

What do you think @jajourda ?

**Comment by jajourda** [comment]
@Vadorequest , honestly, it's a tossup. 

when I hear "preset" i think of the concepts of **_approach_** and/or  **_methodology_**.  When I hear "variant" I think of **_framework_** or **_solution_**. I think Next-right-now is a framework/solution more than an approach.

I prefer **variants**, as it seems the most fitting noun that describes the nature of the feature. But I defer to your awareness of trending terms.

**Comment by jajourda** [comment]
##### on second thought...

i think preset is possibly better and more in line with present usage.

If you are conceptualizing to Next-right-now as a **_configuration_**, i can employing preset as being better in line with how "preset" is currently being used across many repos on github.

**Comment by Vadorequest** [comment]
Thanks for your feedbacks ;) 

I do think about it as a configuration, and I do think people may understand it better if using a familiar word, rather than one I kinda feel like I invented on the fly.

**Comment by Vadorequest** [comment]
Implemented through https://github.com/UnlyEd/next-right-now/pull/22 (documentation)

Next step is to add a bunch of new presets: https://unlyed.github.io/next-right-now/concepts/presets#which-presets-are-being-considered

**Comment by Vadorequest** [comment]
RFC implementation is over. 

Also, the main github branch is now `gh-pages` instead of `master`, which has been deprecated (kept for external links support, avoid breaking external links)

All doc has been moved to gh-pages branch, and removed from other branches.

This was tedious to do and quite time consuming, thanks for your inputs and feedback @jajourda !

New presets will be created soon. I'm gonna try out Directus to replace GraphCMS as an OSS (and free) alternative.

**Comment by jajourda** [comment]
@Vadorequest , i 'm excited to see directus in action! 

### thank you for all your  hard work on the docs!
